stl/src/vector_algorithms.cpp

```diff
@@ -2930,10 +2930,378 @@ namespace {
+                const unsigned int _Bingo = _mm256_movemask_ps(_mm256_castsi256_ps(_Mask));
+                if (_Bingo != 0) {
+                    return _Ix + _tzcnt_u32(_Bingo);
+                }
+            }
+
+            const size_t _Haystack_length_tail = _Haystack_length & 7;
+            if (_Haystack_length_tail != 0) {
+                const unsigned int _Tail_bingo_mask = (1 << _Haystack_length_tail) - 1;
+                _Ty _Buf[8];
+                memcpy(_Buf, _Haystack_ptr + _Haystack_length_vec, _Haystack_length_tail * sizeof(_Ty));
+                const __m256i _Data       = _Load_avx_256_8(_Buf);
+                const __m256i _Mask       = _Mask_out_oveflow<_Ty>(_Step(_Bitmap, _Data), _Data);
+                const unsigned int _Bingo = _mm256_movemask_ps(_mm256_castsi256_ps(_Mask)) & _Tail_bingo_mask;
+                if (_Bingo != 0) {
+                    return _Haystack_length_vec + _tzcnt_u32(_Bingo);
+                }
+            }
+
+            return static_cast<size_t>(-1);
+        }
+
+        template <class _Ty>
+        size_t _Impl_last_avx(const void* const _Haystack, size_t _Haystack_length, const void* const _Needle,
+            const size_t _Needle_length) noexcept {
+            const auto _Haystack_ptr = static_cast<const _Ty*>(_Haystack);
+            const auto _Needle_ptr   = static_cast<const _Ty*>(_Needle);
+
+            const __m256i _Bitmap = _Make_bitmap(_Needle_ptr, _Needle_length);
+
+            while (_Haystack_length >= 8) {
+                _Haystack_length -= 8;
+                const __m256i _Data       = _Load_avx_256_8(_Haystack_ptr + _Haystack_length);
+                const __m256i _Mask       = _Mask_out_oveflow<_Ty>(_Step(_Bitmap, _Data), _Data);
+                const unsigned int _Bingo = _mm256_movemask_ps(_mm256_castsi256_ps(_Mask));
+                if (_Bingo != 0) {
+                    return _Haystack_length + 31 - _lzcnt_u32(_Bingo);
+                }
+            }
+
+            const size_t _Haystack_length_tail = _Haystack_length & 7;
+            if (_Haystack_length_tail != 0) {
+                const unsigned int _Tail_bingo_mask = (1 << _Haystack_length_tail) - 1;
+                _Ty _Buf[8];
+                memcpy(_Buf, _Haystack_ptr, _Haystack_length_tail * sizeof(_Ty));
+                const __m256i _Data       = _Load_avx_256_8(_Buf);
+                const __m256i _Mask       = _Mask_out_oveflow<_Ty>(_Step(_Bitmap, _Data), _Data);
+                const unsigned int _Bingo = _mm256_movemask_ps(_mm256_castsi256_ps(_Mask)) & _Tail_bingo_mask;
+                if (_Bingo != 0) {
+                    return 31 - _lzcnt_u32(_Bingo);
+                }
+            }
+
+            return static_cast<size_t>(-1);
+        }
+#endif // ! _M_ARM64EC
+
+        using _Scalar_table_t = bool[256];
+
+        template <class _Ty>
+        bool _Build_scalar_table(const void* const _Needle, const size_t _Needle_length, bool* _Table) noexcept {

```

Comment:
Thanks, I learned something new about ABI.

I expected that reference to an array would be a pointer to a pointer instead of just pointer, since just an array is a pointer, so this would be a slight pessimization. But it isn't, reference to an array is just a pointer, no pessimization.